Added evaluation of variable expressions

ggetz opened the pull request:
Added evaluation of variable expressions.

I added a helper function to take all occurrences of `${name}` and translate them to `czm_name` before running the expression through jsep. 

Currently, if the feature does not have that property name, it returns `undefined`. I think this behavior is better than throwing an error in case of data sets that don't uniformly have every property defined for each feature.

Also updated the specs.

pjcozzi commented:
> Currently, if the feature does not have that property name, it returns undefined. I think this behavior is better than throwing an error in case of data sets that don't uniformly have every property defined for each feature.

Yes, this is the right approach.

pjcozzi commented:
How do strings work, e.g.,
```
${name} === 'a-name'
```
or
```
'${name}' === 'a-name'
```

Likewise:
```
${another-name} === 1
```
and perhaps
```
'${another-name}' === '1'
```

pjcozzi commented on Source/Scene/Expression.js:
```diff
@@ -52,6 +52,21 @@ define([
         setEvaluateFunction(this);
     }
 
+    function replaceVariables(expression) {
```
I'm the wrong perhaps to ask, but I suspect a one-liner regex could do this.  @tfili would know.

pjcozzi replied:
I still think we should consider a regex here.  Can you ask @tfili how hard it would be?  I suspect it is easy.

The other thing we can consider is modifying `jsep` to allow for `${` and `}` to indicate variables, just like it can already parse functions.  This could probably be done in an elegant way that lets us add new notation later.

mramato replied:
Is this just replacing all instances of `${varName}` with the value of the varName property?  You can't do it with a single regex, but yes, it can be much much simpler.

Do we account for people that actually want `${varName}` in a string?  Can you escape the `$` with `\`?

ggetz replied:
OK, can you point me in the direction of the simpler way to implement this?

Also I can add an escape character.

pjcozzi commented on Specs/Scene/ExpressionSpec.js:
```diff
@@ -13,6 +13,33 @@ defineSuite([
+
+    it('evaluates variable', function() {
+        var feature = new MockFeature();
+        feature.addProperty('height', 10);
```
Add tests for all the different types: Color, String, undefined, null, Boolean.

ggetz replied:
Would the color stored in the property be a string, (ie. '#ffffff`) or a Color object?

pjcozzi replied:
See how it works out with Cesium's `Color` object.  We can add `addColorProperty` if we need to, but we might just have to do some `typeof` checks to handle colors cleanly from the API perspective.

ggetz commented:
Currently the variable only evaluates `${name} === 'a-name'` not `'${name}' === 'a-name'`, as I only replace the variable name with czm_name, and then wait to evaluate the variable until runtime, because that's the only time the feature is available to evaluate the property. 

I also released I did not account for unmatched curly braces. I will throw an error for that.

ggetz commented:
Should we evaluate variable name inside strings at all? We cannot do that during the initial parsing of the expression, because we do not have the feature available to evaluate it. I could also find and replace the variable names inside the string during runtime, but that seems bad performance-wise to run that operation repeatedly on every string being evaluated.

pjcozzi commented:
> Currently the variable only evaluates ${name} === 'a-name' not '${name}' === 'a-name'

When we do non-visual properties I can see something like `'${name}, ${description}'` being used often so I suggest we add support now.  I will also think through this more carefully when I do more spec work tomorrow.

pjcozzi commented:
> Should we evaluate variable name inside strings at all? We cannot do that during the initial parsing of the expression, because we do not have the feature available to evaluate it.

Yes, we would flag the node at compile time, so we know we need to do the replacement at runtime.  It won't be blazingly fast, but we can optimize quite a bit to minimize string searching and copies.

ggetz commented:
> Yes, we would flag the node at compile time, so we know we need to do the replacement at runtime. It won't be blazingly fast, but we can optimize quite a bit to minimize string searching and copies.

OK

ggetz commented:
> When we do non-visual properties I can see something like '${name}, ${description}' being used often so I suggest we add support now. I will also think through this more carefully when I do more spec work tomorrow.

OK, I implemented this. I added a new `ExpressionNodeType`, `VARIABLE_STRING`. I check each string at compile time to see if it contains `${` and flags the type as `VARIABLE_STRING`, and set the evaluation function so this is evaluated at runtime. 

Strings that have '${` but not the closing `}` are just left as is, they do not throw an error.

For replacing colors in strings, I use the default `Color.toString()` function.

I added tests to check for each type, as well as each type being replaced inside variable strings.

pjcozzi commented on Source/Scene/ExpressionNodeType.js:
```diff
@@ -17,7 +17,8 @@ define([
         LITERAL_STRING: 7,
-        LITERAL_COLOR: 8
+        LITERAL_COLOR: 8,
+        VARIABLE_STRING : 9
```
Minor, but maybe `VARIABLE_IN_STRING` is more precise.

pjcozzi commented on Source/Scene/Expression.js:
```diff
@@ -226,6 +281,33 @@ define([
+    Node.prototype._evaluateVariableString = function(feature) {
+        var exp = this._value;
+        var result = "";
+        var i = exp.indexOf('${');
```
We can optimize this by breaking up the string at runtime based on where the variables are, and then just concatenating here.  Basically, cache the `indexOf` calls at compile time.

We could even go farther to avoid memory allocations here (which will be implicit in the concatenating) but it let's wait until we profile since I don't expect these type of expressions to get evaluated often (most likely for UI, mouse over, etc.).

mramato replied:
There is no need to iterate the string manually at all.  This can be done with a couple of regexes and string.replace. My comment about escaped `$` also applies here.

pjcozzi replied:
Not sure that I would try the regex engine performance here, but could be fine since I don't expect it to be a hotspot for common use cases.

mramato replied:
Regex performance is actually way faster than anything we would do manually.

mramato replied:
Here's one approach using regex (I haven't actually tested it yet but it should work).  I think this does what you want. I'm sure it can be further optimized and cleaned up.  It also doesn't check for `\$` yet, but that should be doable.

@ggetz I'll stop by to walk you through this.

```javascript
var variablesRegex = /\${(.*?)}/g;

var result = expression;
var matches = variablesRegex.exec(result);
while(definedNotNull(matches)) {
	var placeholder = matches[0];
	var variableName = matches[1];
	var property = feature.getProperty(variableName);
	result = result.replace(placeholder, property);
	match = variablesRegex.match(result);
}
```

ggetz replied:
Awesome, thanks! I was implementing something pretty similar, but this is a little cleaner.

ggetz replied:
I think `/[^\\]\${(.*?)}/g` ignores the `\$`

mramato replied:
I don't think that's right because you need to make sure `\$` doesn't appear, not just `\`.  You can use the debugger console to test this easily without before writing a formal test.

On the other hand, (and I could be wrong) but I believe the plan is to implement exactly what [ES6 template strings](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/template_strings) do, so we should use that as a reference and I'm not sure how they let you escape the `$`.  @pjcozzi is my assumption here correct?

pjcozzi commented:
Nice, just those comments.

pjcozzi commented:
When a variable is inside a string, we should allow `${}` to be escaped, e.g., `\${}`.  For example, try
```
let person = {name: 'John Smith'};
let tpl = `My name is ${person.name}.`;

console.log(tpl);
```
and
```
let person = {name: 'John Smith'};
let tpl = `My name is \${person.name}.`;

console.log(tpl);
```
here: http://www.es6fiddle.net/

Also, see 11.8.6.1 of the spec: http://www.ecma-international.org/ecma-262/6.0/

However, when `${}` is not inside a string, escaping it should be a syntax error, it would be like writing:
```
var i = 1;
\i++;
```
If this is painful, we could make it a TODO or come up with another way to declare variables (we should think of styling more as a full programming language than just a string in JSON so perhaps we are overloading `${}` as both a built-in variable and string template.  I dunno; don't want to overthink this now until we get more experience writing styles - then things will become obvious quickly).

pjcozzi commented:
@ggetz merge in `3d-tiles-style`.

ggetz commented:
> When a variable is inside a string, we should allow ${} to be escaped, e.g., \${}.

I'm currently working on this.

ggetz commented:
Well one problem with using the escape characters seems to be that when jsep parses a string, and escape characters are ignored, so the string `'\${name}'` its parsed as `'{name}'`. I guess we could find/replace before running it through jsep? But then we have to worry about escaping the sequence we are replacing it with as well.

pjcozzi commented:
> I guess we could find/replace before running it through jsep? But then we have to worry about escaping the sequence we are replacing it with as well.

Sounds like it would work to me.

ggetz commented:
So this seems to be an issue with jsep, https://github.com/soney/jsep/issues/36, so no matter how many `\` characters we insert, when the string is evaluated it will not contain any `\`.

ggetz commented:
Also [templete strings](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/template_strings) themselves don't seem to have the ability to be escaped.

pjcozzi commented:
@ggetz do whatever you think is reasonable for now and update the TODOs in the roadmap; no need to let this case slow us down.

pjcozzi commented:
Let me know when this is ready.

ggetz commented:
OK, I think I will add this to the TODO roadmap for now. Should I leave in the test to check for escaped variables?

pjcozzi commented:
Just make sure the tests and in-sync with the implementation; no need to have extra or commented out tests.

ggetz commented:
OK, I made the changes you suggested, and used find/replace with regex's to replace the variables in the string. I didn't replace the initial find/replace of the variables with a regex yet, as I couldn't find a regex that worked for only matching occurrences outside of strings.

pjcozzi commented on Source/Scene/Expression.js:
```diff
@@ -278,6 +323,28 @@ define([
 
+    Node.prototype._evaluateVariableString = function(feature) {
+        var result = this._value;
+        var regex = /\${(.*?)}/g;
```
Construct this just once out of this function as a file-scoped variable.

In general, remember that we want to avoid everything that we can at runtime.

I'm still confident that we can make this a lot faster without the regex, but no need to bother now.

pjcozzi commented:
Just that comment.

ggetz commented:
Updated.

pjcozzi commented:
:+1:

pjcozzi commented:
I just realized that for variables in strings (as discussed [above](Template Literal)), back-ticks should be used for the strings: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals

@ggetz does this make the implementation easier or harder?

ggetz commented:
So they should never be found in a regular string denoted with `"` or `'`?

ggetz commented:
That still does not help the escaped characters issue, but makes checking for the variables in the strings simpler.
